app/prepare_proposal.go

```diff
@@ -1,30 +1,92 @@
 	"github.com/cosmos/cosmos-sdk/x/auth/signing"
 	abci "github.com/tendermint/tendermint/abci/types"
 	"github.com/tendermint/tendermint/pkg/consts"
+	"github.com/tendermint/tendermint/pkg/da"
 	core "github.com/tendermint/tendermint/proto/tendermint/types"
-	coretypes "github.com/tendermint/tendermint/types"
 )
 
-// PreprocessTxs fullfills the celestia-core version of the ACBI interface, by
-// performing basic validation for the incoming txs, and by cleanly separating
-// share messages from transactions
+// PrepareProposal fullfills the celestia-core version of the ACBI interface by
+// preparing the proposal block data. The square size is determined by first
+// estimating it via the size of the passed block data. Then the included
+// MsgWirePayForData messages are malleated into MsgPayForData messages by
+// separating the message and transaction that pays for that message. Lastly,
+// this method generates the data root for the proposal block and passes it the
+// blockdata.
 func (app *App) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
-	squareSize := app.SquareSize()
-	shareCounter := uint64(0)
-	var shareMsgs []*core.Message
-	var processedTxs [][]byte
-	for _, rawTx := range req.BlockData.Txs {
+	squareSize := app.estimateSquareSize(req.BlockData)
+
+	dataSquare, data := SplitShares(app.txConfig, squareSize, req.BlockData)
+
+	eds, err := da.ExtendShares(squareSize, dataSquare)
+	if err != nil {
+		app.Logger().Error(
+			"failure to erasure the data square while creating a proposal block",
+			"error",
+			err.Error(),
+		)
+		panic(err)
+	}
+
+	dah := da.NewDataAvailabilityHeader(eds)
+	data.Hash = dah.Hash()
+	data.OriginalSquareSize = squareSize
+
+	return abci.ResponsePrepareProposal{
+		BlockData: data,
+	}
+}
+
+// estimateSquareSize returns an estimate of the needed square size to fit the
+// provided block data. It assumes that every malleatable tx has a viable commit
+// for whatever square size that we end up picking.
+func (app *App) estimateSquareSize(data *core.Data) uint64 {
+	txBytes := 0
+	for _, tx := range data.Txs {
+		txBytes += len(tx) + delimLen(uint64(len(tx)))
+	}
+	txShareEstimate := txBytes / consts.TxShareSize
+	if txBytes > 0 {
+		txShareEstimate++ // add one to round up
+	}
+
+	evdBytes := 0
+	for _, evd := range data.Evidence.Evidence {
+		evdBytes += evd.Size() + delimLen(uint64(evd.Size()))
+	}
+	evdShareEstimate := evdBytes / consts.TxShareSize
+	if evdBytes > 0 {
+		evdShareEstimate++ // add one to round up
+	}
+
+	msgShareEstimate := estimateMsgShares(app.txConfig, data.Txs)
+
+	totalShareEstimate := txShareEstimate + evdShareEstimate + msgShareEstimate
+
+	estimatedSize := types.NextPowerOf2(uint64(math.Sqrt(float64(totalShareEstimate))))
+
+	if estimatedSize > consts.MaxSquareSize {
+		return consts.MaxSquareSize
+	}
+
+	if estimatedSize < consts.MinSquareSize {
+		estimatedSize = consts.MinSquareSize
+	}
+
+	return estimatedSize
```

Comment:
This control flow is a bit weird because it's uneven: the first `if` returns and the second doesn't. If this were Rust you could use a `match`, but given the constraints of the language I would prefer an `if` that sets `estimatedSize` followed by an `else if` following by a `return`.

Reply:
we can use a switch statement, wdyt?

https://github.com/celestiaorg/celestia-app/pull/224/commits/a17beb0e99f9815842a260df74cbcc2261525d33